client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/PingOperation.java

```diff
@@ -74,30 +77,64 @@ public void acceptResponse(ByteBuf buf, short status, HeaderDecoder decoder) {
 
    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-      if (cause instanceof HotRodClientException && cause.getMessage().contains("CacheNotFoundException")) {
-         complete(PingResult.CACHE_DOES_NOT_EXIST);
-         return;
+      while (cause instanceof DecoderException && cause.getCause() != null) {
+         cause = cause.getCause();
+      }
+      PingOperation.PingResponse pingResponse = new PingOperation.PingResponse(cause);
+      if (pingResponse.isCacheNotFound()) {
+         complete(pingResponse);
       } else {
          super.exceptionCaught(ctx, cause);
       }
    }
 
-   public enum PingResult {
-      // Success if the ping request was responded correctly
-      SUCCESS,
-      // Success with compatibility enabled
-      SUCCESS_WITH_COMPAT,
-      // When the ping request fails due to non-existing cache
-      CACHE_DOES_NOT_EXIST,
-      // For any other type of failures
-      FAIL;
+   public static class PingResponse {
+
+      public static PingResponse EMPTY = new PingResponse(null);
+
+      private final short status;
+      private final MediaType keyMediaType;
+      private final MediaType valueMediaType;
+      private final Throwable error;
+
+      PingResponse(short status, MediaType keyMediaType, MediaType valueMediaType) {
+         this.status = status;
+         this.keyMediaType = keyMediaType;
+         this.valueMediaType = valueMediaType;
+         this.error = null;
+      }
+
+      PingResponse(Throwable error) {
+         this.status = -1;
+         this.keyMediaType = MediaType.APPLICATION_UNKNOWN;
+         this.valueMediaType = MediaType.APPLICATION_UNKNOWN;
+         this.error = error;
+      }
+
+      public short getStatus() {
+         return status;
+      }
 
       public boolean isSuccess() {
-         return this == SUCCESS || this == SUCCESS_WITH_COMPAT;
+         return HotRodConstants.isSuccess(status);
+      }
+
+      public boolean isObjectStorage() {
+         boolean storageHinted = status == NO_ERROR_STATUS_OBJ_STORAGE
```

Comment:
I'm not sure I like these checks staying here. They should be moved to Codec28 so that these checks are only done if using HR protocol 2.8 (or whichever is the last HR protocol version that sent them). If you leave them here, you're going to carry checking even with protocol 2.9 which is wasteful since the server does not send them.

Reply:
ok, let me refactor it